File: etc/bench/drive.ml
open Slice
let atom s =
  match float_of_string_opt s with Some f -> N f | None -> S s
let split c s =
  String.split_on_char c s
let () =
  let p = 2.0 and few = 128 and start = 4 and stop = 50 in
  let ic = open_in "data.csv" in
  let ls = ref [] in
  (try while true do
    let l = String.trim (input_line ic) in
    if l <> "" then ls := l :: !ls
  done with End_of_file -> close_in ic);
  let cells = List.map (split ',') (List.rev !ls) in
  let names = List.hd cells in
  let rs = List.map (fun r -> Array.of_list (List.map atom r)) (List.tl cells) in
  let t = mktbl names rs in
  Printf.printf "n %d x [%s] y [%s]\n" (List.length t.rows)
    (String.concat ", " (List.map string_of_int t.xat))
    (String.concat ", " (List.map (fun (a,w) ->
       Printf.sprintf "(%d, %d)" a (int_of_float w)) t.yat));
  Printf.printf "ymu %.6f\n" (ymu p t t.rows);
  let rv = Array.of_list t.rows in
  Printf.printf "ydist0 %.6f\n" (ydist p t rv.(0));
  Printf.printf "xdist01 %.6f\n" (xdistr p t rv.(0) rv.(1));
  Printf.printf "div %s\n"
    (String.concat " " (List.map (fun (_,c) -> Printf.sprintf "%.6f" (divc c)) t.cols));
  Printf.printf "mids %s\n"
    (String.concat " " (List.map (fun (a,v) ->
       Printf.sprintf "%d:%.6f" a (numf v)) (mids t)));
  let ic2 = open_in "gauss.txt" in
  let gl = ref [] in
  (try while true do gl := input_line ic2 :: !gl done with End_of_file -> close_in ic2);
  let g = Array.of_list (List.rev_map (fun l ->
    List.map float_of_string (split ',' (String.trim l))) !gl) in
  Printf.printf "same %d %d\n"
    (if same g.(0) g.(1) 0.0 then 1 else 0)
    (if same g.(0) g.(2) 0.0 then 1 else 0);
  let t0 = Unix.gettimeofday () in
  let r1 = int_of_string Sys.argv.(1) and r2 = int_of_string Sys.argv.(2) in
  let s = ref 0.0 in
  for _ = 1 to r1 do
    for i = 0 to 119 do
      for j = i + 1 to 119 do s := !s +. xdistr p t rv.(i) rv.(j) done
    done
  done;
  Printf.printf "W1 %.6f\n" !s;
  let s2 = ref 0.0 in
  for _ = 1 to r2 do
    s2 := !s2 +. ydist p t (List.hd (acquire p few start stop t))
  done;
  Printf.printf "W2 %.6f\n" !s2;
  Printf.printf "SEED %d\n" !seed;
  Printf.eprintf "secs %.3f\n" (Unix.gettimeofday () -. t0)

File: etc/bench/drive.lua
P, Few, Start, Stop, SEED = 2, 128, 4, 50, 1
dofile("slice.lua")
local function atom(s) return tonumber(s) or s end
local function split(s,    t) t={}
  for w in s:gmatch("[^,]+") do t[#t+1]=w end
  return t end
local raw = {}
for line in io.lines("data.csv") do
  if line:match("%S") then raw[#raw+1] = split((line:gsub("%s+$",""))) end end
local src = {raw[1]}
for i=2,#raw do local r={}
  for j=1,#raw[i] do r[j-1] = atom(raw[i][j]) end
  src[#src+1] = r end
local t = Tbl(src)
local R = t.rows
local xs = {}
for _,a in ipairs(t.x) do xs[#xs+1]=a end
local ys = {}
for _,aw in ipairs(t.y) do ys[#ys+1] = string.format("(%d, %d)", aw[1], aw[2]) end
print(string.format("n %d x [%s] y [%s]", #R, table.concat(xs,", "), table.concat(ys,", ")))
print(string.format("ymu %.6f", ymu(t,R)))
print(string.format("ydist0 %.6f", ydist(t,R[1])))
print(string.format("xdist01 %.6f", xdist(t,R[1],R[2])))
local dv = {}
for at=0,#t.names-1 do if t.cols[at] then dv[#dv+1]=string.format("%.6f", div(t.cols[at])) end end
print("div "..table.concat(dv," "))
local md = {}
for _,at in ipairs(t.x) do md[#md+1]=string.format("%d:%.6f", at, mids(t)[at]) end
print("mids "..table.concat(md," "))
local g = {}
for line in io.lines("gauss.txt") do
  local row={} ; for w in line:gmatch("[^,]+") do row[#row+1]=tonumber(w) end
  g[#g+1]=row end
print(string.format("same %d %d", same(g[1],g[2]) and 1 or 0, same(g[1],g[3]) and 1 or 0))
local t0 = os.clock()
local s = 0.0
for _=1,tonumber(arg[1]) do
  for i=1,120 do for j=i+1,120 do s = s + xdist(t,R[i],R[j]) end end end
print(string.format("W1 %.6f", s))
local s2 = 0.0
for _=1,tonumber(arg[2]) do s2 = s2 + ydist(t, acquire(t)[1]) end
print(string.format("W2 %.6f", s2))
print("SEED "..math.floor(SEED))
io.stderr:write(string.format("secs %.3f\n", os.clock()-t0))
